### PR TITLE
Give God mode algorithm more exposure and proper test coverage #4624

### LINKED ISSUE #4378: Remove redundant regex replace in processPageSourceForGodMode() in AppPage.java
this line in AppPage.java (line 791) is no longer needed:
```
.replaceAll(Const.ParamsNames.REGKEY + "%3D([a-zA-Z0-9]){1,}\\%", Const.ParamsNames.REGKEY + "%3D{*}\\%")
```

### LINKED ISSUE #4624: Give God mode algorithm more exposure and proper test coverage
Related to #4623. If we're going to make God mode mainstream, the code shouldn't be obscured in `AppPage` anymore. We should move it to `HtmlHelper` and give it a proper coverage.
`GodModeTest` does provide the test (albeit outdated), but is only run occasionally while the algorithm has been part of the main testing process ever since we abandoned the `{*}` wildcards and adopted `${placeholder}`.

### PR DESCRIPTION
**Pull request by wkurniawan07**
Fixes #4624 
I've checked every single replace line on the algorithm and removed all that's unused while migrating, so this also fixes #4378.

### CODE REVIEW

File: src/test/java/teammates/test/driver/HtmlHelper.java [hunk 1]
```diff
@@ -287,4 +304,143 @@ private static boolean isVoidElement(String elementName){
...
+                      .replace(TP.TEST_STUDENT1_ACCOUNT, "${test.student1}")
+                      .replace(TP.TEST_STUDENT2_ACCOUNT, "${test.student2}")
+                      .replace(TP.TEST_INSTRUCTOR_ACCOUNT, "${test.instructor}")
+                      .replace(TP.TEST_ADMIN_ACCOUNT, "${test.admin}");
+    }
+    
+    /**
+     * This method is only used for testing.
+     */
+    public static String injectImportantValuesForTest(String content) {
```
Comment (damithc) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: `Important` is a bit subjective. Any way to give a better name? e.g. `ContextDependent`

File: src/test/java/teammates/test/pageobjects/AppPage.java [hunk 2]
```diff
@@ -812,104 +797,6 @@ private boolean testAndRunGodMode(String filePath, String content, boolean isPar
...
-        return content
-                // this handles the logout url that google generates
-                .replaceAll("_ah/logout\\?continue=.*?\"", "_ah/logout?continue=\\${continue\\.url}\"")
-                // photo from instructor
-                .replaceAll(Const.ActionURIs.STUDENT_PROFILE_PICTURE + "\\?" + Const.ParamsNames.STUDENT_EMAIL + "=([a-zA-Z0-9]){1,}\\&amp;"
-                        + Const.ParamsNames.COURSE_ID + "=([a-zA-Z0-9]){1,}", 
-                        Const.ActionURIs.STUDENT_PROFILE_PICTURE + "\\?" + Const.ParamsNames.STUDENT_EMAIL 
-                        + "=\\${student\\.email\\.enc}\\&amp;" + Const.ParamsNames.COURSE_ID + "=\\${course\\.id\\.enc}")
-                .replaceAll(Const.ActionURIs.STUDENT_PROFILE_PICTURE + "\\?" + Const.ParamsNames.COURSE_ID + "=([a-zA-Z0-9]){1,}\\&amp;"
-                        + Const.ParamsNames.STUDENT_EMAIL + "=([a-zA-Z0-9]){1,}",
```
Comment (thyageshm) [inline src/test/java/teammates/test/pageobjects/AppPage.java]: @wkurniawan07, I think we had courseid followed by studentemail and studentemail followed by courseid because both can occur depending on the browser being used

File: src/test/resources/pages/godmode.html [hunk 3]
```diff
@@ -6,10 +6,7 @@
     <div id="mainContent">
         <input type="text" value="">
         <a class="student-profile-pic-view-link btn-link" 
-           data-link="/page/studentProfilePic?studentemail=1CAAAF9A82CEF76B0BA249598DDF9CBC2031E84F1AF45D607A84A83D181C168E&courseid=FA95864E467A11325349CB53B4C6D3C39398C035D135466658ABD102B08D4487">
-           View Photo</a>
-        <a class="student-profile-pic-view-link btn-link" 
-           data-link="/page/studentProfilePic?courseid=FA95864E467A11325349CB53B4C6D3C39398C035D135466658ABD102B08D4487&studentemail=1CAAAF9A82CEF76B0BA249598DDF9CBC2031E84F1AF45D607A84A83D181C168E">
```
Comment (thyageshm) [inline src/test/resources/pages/godmode.html]: This test case was there to check the reversal in parameters in some browsers. I think chrome/firefox have different ways to order the parameters.

File: src/test/java/teammates/test/cases/testdriver/HtmlHelperTest.java [hunk 4]
```diff
@@ -135,4 +135,20 @@ public void testConvertToStandardHtmlPart() throws Exception {
         
         //other cases are tested in testComparison
     }
+    
+    @Test
+    public void testReplacement() throws IOException {
```
Comment (thyageshm) [inline src/test/java/teammates/test/cases/testdriver/HtmlHelperTest.java]: If we still have GodModeTest, why have this?

Comment (thyageshm) [inline src/test/java/teammates/test/cases/testdriver/HtmlHelperTest.java]: Ok I see... if this is a lower level testing of the godmode feature, then we have to test after every function call. Like check if injectContextDependentValuesForTest worked properly AND if processPageSourceForHtmlComparison worked inependently

**Comment by thyageshm** [comment]
I think this would require an update of the GodMode docs that you had written previously

File: src/test/java/teammates/test/pageobjects/AppPage.java (hunk 2 above)
Comment (wkurniawan07) [inline src/test/java/teammates/test/pageobjects/AppPage.java]: I have to disagree. If you do a workspace-wide search you won't find any occurrence of the former.

File: src/test/java/teammates/test/cases/testdriver/HtmlHelperTest.java (hunk 4 above)
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/testdriver/HtmlHelperTest.java]: That wasn't the intention actually... this is meant to be the test for processes gone through while doing 'verifyHtml*`. `GodModeTest` is testing God mode, i.e regeneration. They have a lot of overlaps, but definitely not the same thing.

**Comment by wkurniawan07** [comment]
Eh? There's no God mode docs. I've tried written a draft before, but haven't published it yet.

**Comment by thyageshm** [comment]
I see, you had created some explanation notes as part of a previous issue. Do they need to be updated?

File: src/test/java/teammates/test/pageobjects/AppPage.java (hunk 2 above)
Comment (thyageshm) [inline src/test/java/teammates/test/pageobjects/AppPage.java]: I see ok

File: src/test/resources/pages/godmode.html (hunk 3 above)
Comment (thyageshm) [inline src/test/resources/pages/godmode.html]: so I guess this is no longer necessary

File: src/test/java/teammates/test/driver/HtmlHelper.java [hunk 5]
```diff
@@ -287,4 +303,141 @@ private static boolean isVoidElement(String elementName){
...
+     */
+    public static String processPageSourceForHtmlComparison(String content) {
+        return replaceUnpredictableValuesWithPlaceholders(
+                      suppressVariationsInInjectedValues(content));
+    }
+
+    /**
+     * Processes the string from web page source for regeneration of expected HTML.
+     */
+    public static String processPageSourceForExpectedHtmlRegeneration(String content, boolean isPart) {
```
Comment (thyageshm) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: I think we should specify the precondition that "content" here should have been through the ```replaceUnpredictableValuesWithPlaceholders``` function

File: src/test/java/teammates/test/driver/HtmlHelper.java [hunk 6]
```diff
@@ -287,4 +303,141 @@ private static boolean isVoidElement(String elementName){
...
+        return content // replace dev server admin absolute URLs (/_ah/...) with their relative counterparts
+                      .replace("\"" + TP.TEAMMATES_URL + "/_ah", "\"/_ah")
+                      // replace truncated long accounts with their original counterparts
+                      .replace(StringHelper.truncateLongId(TP.TEST_STUDENT1_ACCOUNT), TP.TEST_STUDENT1_ACCOUNT)
+                      .replace(StringHelper.truncateLongId(TP.TEST_STUDENT2_ACCOUNT), TP.TEST_STUDENT2_ACCOUNT)
+                      .replace(StringHelper.truncateLongId(TP.TEST_INSTRUCTOR_ACCOUNT), TP.TEST_INSTRUCTOR_ACCOUNT)
+                      .replace(StringHelper.truncateLongId(TP.TEST_ADMIN_ACCOUNT), TP.TEST_ADMIN_ACCOUNT);
+    }
+
+    private static String replaceUnpredictableValuesWithPlaceholders(String content) {
```
Comment (thyageshm) [inline src/test/java/teammates/test/driver/HtmlHelper.java]: A little explanation as to what this function does would be helpful. Eg: "Substitutes values that are different across various test runs with placeholders. These values are identified using their known format"

File: src/test/java/teammates/test/cases/testdriver/HtmlHelperTest.java (hunk 4 above)
Comment (thyageshm) [inline src/test/java/teammates/test/cases/testdriver/HtmlHelperTest.java]: I dont understand. Are we testing the lower level functions that consitute Godmode over here? If yes, I think this test should be a little more stringent. Eg: Check the expected string after line 142, 143 and 146. The ```assertSameHtml``` statement in line 147 can be removed

**Comment by thyageshm** [comment]
Just a few more comments. After this it is good to be merged

**Comment by wkurniawan07** [comment]
If you're talking about the gist I wrote long ago (who still read it now, btw?), then no, because no part of the process is modified. This PR only moves the relevant process to `HtmlHelper` and gives test coverage. I did remove the unused replace lines, but the internal workings of the algorithm was never a point of discussion in the gist.

File: src/test/java/teammates/test/cases/testdriver/HtmlHelperTest.java (hunk 4 above)
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/testdriver/HtmlHelperTest.java]: No, you misunderstood it entirely. Here what is being tested is the process of HTML verification, i.e the replacement algorithms. Copy-pasting from the gist I wrote:

- Expected string, from test file → inject test properties → convert to standard HTML
- Actual string, from web page → suppress variations of injected values, replace non-injected values with wildcards/placeholders → convert to standard HTML

Now take a look at how the `expected` and `actual` are processed in the test and you'll see that they match the above steps. That's exactly what's being tested. This test has no knowledge of this thing called "God mode" whatsoever.

Comment (damithc) [inline src/test/java/teammates/test/cases/testdriver/HtmlHelperTest.java]: On a related note, try to migrate relevant content in the Gist to code comments and/or dev docs. Can be a separate issue.

Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/testdriver/HtmlHelperTest.java]: Either option doesn't sound good; whatever I've written in the Gists before (for reference, [this](https://gist.github.com/wkurniawan07/e1fb00be615bb34ca01f) and [this](https://gist.github.com/wkurniawan07/d5d7dc842fdb922bbcb9)) are quite detailed and technical, and would be against the "minimal yet sufficient documentation" principle. A technical report might be a better alternative.

Comment (damithc) [inline src/test/java/teammates/test/cases/testdriver/HtmlHelperTest.java]: I'm asking to ensure 'minimal and sufficient documentation' for god mode. It doesn't have to be as detailed as the gist. Dev docs need to describe how to use God mode. Code comments should be enough for future maintainers of God mode code. I consider Gists as temporary notes we created to facilitate the discussion, and as such, not part of long-term documentation.

Comment (damithc) [inline src/test/java/teammates/test/cases/testdriver/HtmlHelperTest.java]: @wkurniawan07 and @thyageshm, the two of you can cross check to ensure:
a) dev docs (where we explain process) explains usage of god mode
b) code comments or dev docs (where we explain design) explains the design/implementation details needed to maintain god mode. For this kind of documentation, code comments preferred, but some high level info may have to be put in dev docs.

Comment (thyageshm) [inline src/test/java/teammates/test/cases/testdriver/HtmlHelperTest.java]: @wkurniawan07, I still dont get the difference you are talking about. Over here you have manually done the html verification process as opposed to calling ```verifyHtml```. So unless we test after each of the individual steps in the process, this test adds no value (other than the fact that it would be run more often). I think the test at line 97 in GodModeTest.java accomplishes what this test is doing.

Unless you are saying that you are testing the "process" of html verification by checking whether the steps actually work. In that case, it is not really needed as a call to verifyHtml would do exactly that.

Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/testdriver/HtmlHelperTest.java]: Ok, it's good that the issue at hand is understood better now.

1. Yes, it's true that `verifyHtml` does all the steps in one go, however do remember that `verifyHtml` works _with a real web page seen on a browser_ and already has an abstraction on which part of the HTML page source is examined (mainContent, chosen part, entire file, ...). This test, on the other hand, is purely to test how the injection/replacement of values work.
2. With regards to "this test adds no value (other than the fact that it would be run more often)", that's exactly the value I'm looking for, and back to the second reason I mentioned: this algorithm is used _at all times_ instead of only during regeneration, and something that is used _at all times_ should be tested _at all times_ instead of only occasionally (real use case: if we suspect the algorithm to be broken, would you rather run `GodModeTest` just for the sake of checking if it is really broken, or having `HtmlHelperTest` catching it concurrently with the whole test suite?).

In other words, the main purpose of each test:
`HtmlHelperTest`: testing the injection and replacement of values
`GodModeTest`: testing the regeneration of HTML file

They are very different. However, I won't deny that in the current state `GodModeTest` also serves the purpose of `HtmlHelperTest` mentioned. I will agree that this test is no longer needed if we can integrate `GodModeTest` back to the full test suite.

Comment (thyageshm) [inline src/test/java/teammates/test/cases/testdriver/HtmlHelperTest.java]: good explanation :-) 

Quoting from above:
In other words, the main purpose of each test:
```HtmlHelperTest```: testing the injection and replacement of values
```GodModeTest```: testing the regeneration of HTML file

If the htmlhelper test is to test whether the injection and the replacement of values work properly, shouldn't they be tested individually too? combining them weakens the test, doesn't it? Also, I think we should specify this goal of "testing injection and replacement of values" in comments so that later devs dont remove this test thinking it is a repetition.

Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/testdriver/HtmlHelperTest.java]: Not really, because there are only 2 such processes being tested and they're targeting 2 different files (actual and expected). Also, there's not much point in testing after each replace. Let's say we test after only `HtmlHelper.injectTestProperties`. What are we comparing against? It won't pass the `HtmlHelper.assertSameHtml` because the other one is not yet processed, and strictly only after both the actual and expected are processed then the test will pass. We can check for (non-)presence of pattern/placeholder, but exhausting every single of the existing pattern/placeholder will be cumbersome. Combining both processes saves us all these hassles with no real drawback.
I don't think the comment is necessary either. I will revoke my previous opinion on "`GodModeTest` also serves the purpose of `HtmlHelperTest` mentioned", because the usage in `GodModeTest` is not _a test for_, but _an application of_ the HTML verification. If `GodModeTest` is a test for such, then the same can be said for _every single HTML verification_ in existence.

Comment (thyageshm) [inline src/test/java/teammates/test/cases/testdriver/HtmlHelperTest.java]: My only concern is that we are testing right at the end, after processing both the expected and actual html. If we can do one more check (for instance after the ```injectTestProperties```), it strengthens the test we are conducting. This check can be another html compare (just create the html file after injectTestProperties). Since we are conducting the test **without** any processing of the "expected" html, the test is more robust.

I understand that it is pointless testing ```injectContextDependentValuesForTest``` as that function was made for this purpose!

Bottom Line: We are testing whether the output of the processing on both the sides yields the same result. BUT we are not checking whether this result is the CORRECT result. They can be the same, but incorrect.

Comment (damithc) [inline src/test/java/teammates/test/cases/testdriver/HtmlHelperTest.java]: I agree that the test can be stronger. If there is a significant risk a bug slipping through and compromising the entire test suite due to the weakness of this test, then we should strengthen it. If there is no such risk, then we can simply leave it as a TODO to strengthen it later. On a related note, God Mode is a complicated yet vital piece of our infrastructure. So may be when the code is fully cleaned up we should review it on its entirety. @wkurniawan07 @thyageshm and I can do it together. I wonder how to use PRs to do a code review on an existing piece of code :-p

Comment (thyageshm) [inline src/test/java/teammates/test/cases/testdriver/HtmlHelperTest.java]: There is no major risk, so I think it can be left as a TODO :-)

Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/testdriver/HtmlHelperTest.java]: The risk is minor to non-existent, I believe. I don't agree on "we are not checking whether this result is the CORRECT result". Let's take a look at how the different types of random values are handled:

1. Unpredictable values: Processed only in actual; the one in expected is not touched because it is hard-coded to represent what is CORRECT.
2. Injected values: Processed only in expected; the one in actual is not touched because it is hard-coded (or, more aptly, hard-injected) to represent what is CORRECT. (we really don't need to test `HtmlHelper.injectContextDependentValuesForTest`, right?)
3. Variations (currently only truncated IDs): This might be the problematic one, as it's processed both in actual (suppressing the variation) and expected (injecting the standardized values). I agree that this might warrant a correctness check, but if we look at how it is done currently, omitting won't make it that much weaker (in expected, the values are injected to the same placeholders `${test.account}` as in (2), so if (2) is correct then it's safe to conclude that this will also be).

i.e I don't think adding a HTML comparison after one process makes it _that_ stronger (if any stronger, that is), and one sure thing is it adds the hassle of having to modify one extra HTML file if we modify the algorithm. I might change my mind on (3) if we have a rule that really warrant a correctness check, but at the current state, I see little to no need to.

Comment (thyageshm) [inline src/test/java/teammates/test/cases/testdriver/HtmlHelperTest.java]: Lets merge this soon. Everything else is good :-)
